src/protocol/sort/generate_sort_permutations.rs

```diff
@@ -0,0 +1,131 @@
+use embed_doc_image::embed_doc_image;
+
+/// This is an implementation of `GenPerm` (Algorithm 6) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+#[derive(Debug)]
+pub struct GenerateSortPermutations<'a> {
+    input: &'a [u64],
+}
+
+impl<'a> GenerateSortPermutations<'a> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a [u64]) -> GenerateSortPermutations {
+        Self { input }
+    }
+
+    #[allow(dead_code)]
+    #[embed_doc_image("semi_honest_sort", "images/sort/semi-honest-sort.png")]
+    /// This protocol generates permutation of a stable sort for the given inputs.
+    /// ![Generate sort permutations steps][semi_honest_sort]
+    /// Steps
+    /// 1. Obtain bit-wise shares in Field by calling modulus conversion on the input match keys.
+    /// 2. Obtain bit permutation to sort 0th input bit. This is also 0th bit composition (i.e. sigma)
+    /// 3. For 1st until n bits, following steps are repeated
+    /// 3i.  Apply inverse of i-1th bit permutation on ith input bits
+    /// 3ii. Obtain bit permutation to sort ith input bit
+    /// 3iii.Compose ith bit permutation on i-1th composition. We have now obtained ith composition
+    /// 4. Return nth composition as the permutation to sort the given inputs
+    pub async fn execute<F: Field>(
+        &self,
+        ctx: ProtocolContext<'_, F>,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let mut bits = convert_all_shares(&ctx, self.input).await?;
+        let mut sigma = BitPermutations::new(&bits[0])
+            .execute(ctx.narrow(&BitPermutation(0)))
+            .await?;
+        for (i, ith_bit) in bits.iter_mut().enumerate().skip(1) {
+            SecureApplyInv::execute(
+                &ctx.narrow(&ApplyInv(i.try_into().unwrap())),
+                ith_bit,
+                &mut sigma.clone(),
+            )
+            .await?;
+            let mut rho = BitPermutations::new(ith_bit)
+                .execute(ctx.narrow(&BitPermutation(i.try_into().unwrap())))
+                .await?;
+            Compose::new(&mut sigma, &mut rho)
+                .execute(ctx.narrow(&ComposeStep(i.try_into().unwrap())))
+                .await?;
+            sigma = rho;
+        }
+        Ok(sigma)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::Rng;
+    use tokio::try_join;
+
+    use crate::{
+        protocol::{sort::generate_sort_permutations::GenerateSortPermutations, QueryId},
+        test_fixture::{make_contexts, make_world, validate_list_of_shares},
+    };
+
+    #[tokio::test]
+    pub async fn generate_sort_permutations() {
+        let world = make_world(QueryId);
+        let [ctx0, ctx1, ctx2] = make_contexts(&world);
+        let mut rng = rand::thread_rng();
+        let mask = (1_u64 << 63) - 1;
+
+        let batchsize = 30;
```

Comment:
After reading the code, I'm worried that there is some confusion between the number of bits and the number of rows. To ensure that we are definitely not mixing up the two, let's make this number MUCH larger than the number of bits. Let's say 1000.

Reply:
Ahh we cant since maximum number of rows can be = Field size (since we secret share permutation in the field). If we are using Fp31 for now, it can be 31 at max for now.

Reply:
Oh that's right... I forgot about that.... Let me create an Fp(Some large prime) class for you to test with!

Reply:
#190